opentelemetry-sdk/src/trace/span.rs

```diff
@@ -198,26 +198,33 @@ impl opentelemetry::trace::Span for Span {
 
 impl Span {
     fn ensure_ended_and_exported(&mut self, timestamp: Option<SystemTime>) {
-        // skip if data has already been exported
-        let mut data = match self.data.take() {
-            Some(data) => data,
-            None => return,
-        };
-
-        let provider = self.tracer.provider();
         // skip if provider has been shut down
-        if provider.is_shutdown() {
+        if self.tracer.provider().is_shutdown() {
```

Comment:
Is there a particular reason to re-read the `provider`?

Reply:
Yes
We need to pass an `&mut Span` to `on_ending`. But if we grab the `provider` at the beginning of the function it's a immutable borrow that lasts for the whole function.

So we need to:
* Borrow the span immutably to read the provider. Not keep the provider around to release the borrow
* Borrow the span mutably for on_ending
* Borrow the span immutably again to grab the provider span processors

Calling to `self.tracer.provider()` is essentially zero cost. It just takes a reference to a field on the `Span`